### DISCUSSION TITLE
Trial towards a chat application with a relay server

### DISCUSSION THREAD

**Discussion by marvellous987** [libp2p/rust-libp2p#2693] [Q&A]
Hi everyone,

I am trying to build a chat application with a relay server using rust-libp2p. However, the example in [here](https://github.com/libp2p/rust-libp2p/blob/master/examples/chat.rs) does not involve any relay server stated in the [hole punching tutorial](https://docs.rs/libp2p/latest/libp2p/tutorials/hole_punching/index.html).

Could anyone give some help on this? Thank you very much!

**Comment by mxinden** [comment]
:wave: 

You can use https://github.com/libp2p/rust-libp2p/blob/master/protocols/relay/examples/relay_v2.rs as the relay server. Consider increasing all limits to not depend on an upgrade to a direct connection (hole punch).

On the client side you need to make sure to wrap the transport:

https://github.com/libp2p/rust-libp2p/blob/0bce7f7fffb0610ec9f9601ca0245a0735d752b1/protocols/dcutr/examples/client.rs#L90

And include a relay client in your `NetworkBehaviour`:

https://github.com/libp2p/rust-libp2p/blob/0bce7f7fffb0610ec9f9601ca0245a0735d752b1/protocols/dcutr/examples/client.rs#L108

Other than that, you can use the standard chat example, just with the additional `/p2p-circuit` `Multiaddr`es.

Hope this is of some help.

**Reply by dariusc93** [reply]
> Consider increasing all limits to not depend on an upgrade to a direct connection (hole punch).

Although this was posted 3 months ago, is there a reason why one should increase the limits to not depend on dcutr?

**Reply by mxinden** [reply]
@dariusc93 depends on your use-case. Hole punching will never be 100% successful, e.g. with symmetric NATs on both ends. By default Circuit Relay v2 enforces limits on the amount of bytes and time a relay connection can use. In the case where hole punching fails, you might want to fall back to the relayed connection only. In such case, bumping these limits might be an option.
